sql/plan/cached_results.go

```diff
@@ -69,6 +69,7 @@ func (n *CachedResults) RowIter(ctx *sql.Context, r sql.Row) (sql.RowIter, error
 }
 
 func (n *CachedResults) Dispose() {
+	n.noCache = true
```

Comment:
In CachedResults.RowIter we have:

```go
	if rows := n.getCachedResults(); rows != nil {
		return sql.RowsToRowIter(rows...), nil
	} else if n.noCache {
		return n.UnaryNode.Child.RowIter(ctx, r)
	} else if n.cacheDone {
		return nil, ErrEmptyCachedResult
	}
```

So this means that if we ever call RowIter a second time for this node, we will not attempt to cache any results but go straight to the child iter, with no caching. This only "works" because HashLookup ignores RowIter and goes straight to the cache on its CachedResults child. It breaks any time we have a lifecycle where we need to call `[RowIter(), Close()]` multiple times on this node. This is too tight a coupling between HashLookup and CachedResults -- would probably be preferable to merge those into a single node that knows how to manage its lifecycle behavior.